### ISSUE TITLE
Explicit best practices for reviewing and commiting?

### ISSUE THREAD

**Issue by JoKeyser** [freedombox/Plinth#964]
Dear @freedombox/reviewers ,
I recently became a reviewer (thanks for your trust!), thus gaining commit-access, and was wondering what the expected practices are.

Of course I have some vague ideas (and made already a mistake in #962), but I would feel more confident if it was made explicit. My brief look into `git log` shows several styles. I propose to write down a few "best practices" into the HACKING file (not calling it "rules" to prevent over-formalizing our collaboration, and leaving room for deviations, if necessary).

My initial guesses, to achieve keep the log understandable, succinct, and verifiable:
* If possible, commit messages should state `MODULE NAME: <insert what changed>`. Multiple lines are allowed e.g. to indicate which issue is closed/fixed by it, or further aspects to the commit.
* Unless trivial (e.g. typo fixes), create a PR to get your work reviewed.
* Before committing a PR, add the line `Reviewed-by: REVIEWER NAME <REVIEWER EMAIL>` to the commit message.
* If a PR contains unnecessary many commits, squash them into one beforehand.
* Sign your commits with gpg (thus avoid GitHub's buttons and do it locally, where your private key is).
* If you review another reviewer's PR, let him/her commit (to preserve gpg signatures).

Open questions:
* There are a few "Signed-off-by" messages in the log; what does that mean?
* Probably rebasing is necessary to avoid "useless" merge commits? (However this strips away gpg-signatures of the original author...)

**Comment by JoKeyser** [comment]
Uhm, and may I ask for some pointers from experienced reviewers what their workflow is? Maybe you use interactive rebasing? Or some Emacs/magit magic? What are common gotcha's?

**Comment by SunilMohanAdapa** [comment]
@JoKeyser, thank you for the initiative to write these practices.

* "Signed-off-by" is for [developer certificate of origin](https://developercertificate.org/).  Also see [Git man page](https://git-scm.com/docs/git-commit#git-commit---signoff).
* Rebasing to avoid merge commits is good.  It keeps the history much more readable for other people to understand, review, bisect and rollback. The reviewer's GPG signature go onto the rebased commits. This somewhat compensates for the loss of original GPG signatures.  Also given that we have to actually verify the signatures on each commit and the contributor may not be in our web of trust (we allow anonymous contributions).  In such cases, the GPG signatures of the reviewers are more important.

Here are some thoughts in addition to yours:
* Reviewers should get their patches reviewed by others.
* Certain operations which are agreed upon before hand do not require review.  These include updating manual and creating new locales.
* Some patches requires knowledge of multiple technologies.  If a reviewer is not familiar with all the technologies involved then the review is expected to review only portions known and ask for further review. In this case, it is nice to let others know what has been reviewed.
* Reviewer is expected to make sure that the work runs as expected by testing the patch throughly. New contributors usually are not familiar with all the impact areas and may not test everything.  It is okay to rely on tests done by trusted contributors by they specify what exact cases have been tested.
* For commit messages we can point contributors to the usual [git commit message guidelines](https://chris.beams.io/posts/git-commit/).
* A reviewer is expected to ensure that the contributor's work:
  * Does not break the current code.  Code base should always be usable state.  We should also strive to keep code base in a release worthy state.
  * Has not security issues.
  * Follows coding standards of the project.
  * Is properly internationalized.
* When merging patches with multiple commits, try to make sure each patch is good one.  Fixes to patches should be squashed into original patch.  Security issues and other serious issues should be avoided in intermediate patches (even if they are fixed in later patches).
* Reviewer may make some minimal changes or obvious changes to the work before merging it.  Telling the contributor and others about the changes would be nice.
* In cases were the changes required are more fundamental or if the contributor is new, the reviewer should try to encourage the contributor to make changes by giving appropriate feedback. This is one of the key ways we mentor new contributors.
* Have your GPG key signed by at least one other reviewers or Debian developers. This is so that signatures made on the commits can be verified.
* For major architectural changes/decisions, consult other in the project before merging.  Use your discretion on which issues are important.
* Be nice to contributors and give them opportunity to learn.
* Have fun reviewing :)

**Comment by JoKeyser** [comment]
Hey Sunil, thanks a lot for this nice and reasonable list. Just to clarify, I got two follow-up questions:

* When you say "Reviewers should get their patches reviewed by others.", you mean the same as my second point above, right? "Others" means other reviewers?
* This signed-off-by seems important, so I guess we should use it more often? Maybe always?

And I guess for my question about practical workflow, everyone should just strive to understand exactly what they're doing :).

**Comment by SunilMohanAdapa** [comment]
On Monday 07 August 2017 02:08 PM, Johannes Keyser wrote:
> Hey Sunil, thanks a lot for this nice and reasonable list. Just to
> clarify, I got two follow-up questions:
> 
>   * When you say "Reviewers should get their patches reviewed by
>     others.", you mean the same as my second point above, right?
>     "Others" means other reviewers?

Yes and yes.  I suppose both of our statements need to be more explicit :)

>   * This signed-off-by seems important, so I guess we should use it more
>     often? Maybe always?

That would be nice.  Apart from these guidelines for reviewers, we
should have one for all contributors.  Perhaps also a standard
contributor agreement (that just says they agree to the right license
like Fedora's).

However, this should not mean that all first patches from new
contributors get rejected first time.

> 
> And I guess for my question about practical workflow, everyone should
> just strive to understand exactly what they're doing :).

Yes :)  Make mistakes and learn policy also applies to reviewers.  So
nothing too serious.

-- 
Sunil

**Comment by JoKeyser** [comment]
Hi @SunilMohanAdapa , thanks for clarifying. I've made PR #967 with a reordered and rephrased version of both our lists, hope I didn't miss anything. I tried to make it shorter, hope nothing is missing. I thought it's better from now on to keep editing that file, seems more efficient than answering in here.

**Comment by SunilMohanAdapa** [comment]
> Uhm, and may I ask for some pointers from experienced reviewers what their workflow is? Maybe you  use interactive rebasing? Or some Emacs/magit magic? What are common gotcha's?

I prefer and recommend Emacs with magit. I do at least the following with it:
- Interactive rebases and rebases of pull requests.
- Automatically sign my commits.
- Automatically add `Reviewed-by` and `Signed-off-by`.
- Verify signatures on commits in log and colorize accordingly.
- Cherry picks.
- Patch diffs and full diffs.
- Very nice hunk editing
- etc.

I have merged #967. This issue can be closed if you think we have achieved the basics of what we set out to do.
